In [1]:
import os

In [2]:
import zipfile

In [ ]:
!wget https://huggingface.co/datasets/Mireu-Lab/CIC-IDS/resolve/main/ZIP/GeneratedLabelledFlows.zip

--2026-09-29 20:34:52--  https://huggingface.co/datasets/Mireu-Lab/CIC-IDS/resolve/main/ZIP/GeneratedLabelledFlows.zip
Resolving huggingface.co (huggingface.co)... 99.86.101.64, 99.86.101.36, 99.86.101.56, ...
Connecting to huggingface.co (huggingface.co)|99.86.101.64|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://us.gcp.cdn.hf.co/xet-bridge-us/650f0ebb57343388e579bf6a/a6120376221bd45484b2dd28f436f4e9c784ca47c8420321808744eb3bc62225?xip=3Nv0ZBJ-aeg&response-content-disposition=inline%3B+filename*%3DUTF-8%27%27GeneratedLabelledFlows.zip%3B+filename%3D%22GeneratedLabelledFlows.zip%22%3B&X-Xet-Cas-Uid=public&response-content-type=application%2Fzip&user_id=public&Expires=1790717692&Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiaHR0cHM6Ly91cy5nY3AuY2RuLmhmLmNvL3hldC1icmlkZ2UtdXMvNjUwZjBlYmI1NzM0MzM4OGU1NzliZjZhL2E2MTIwMzc2MjIxYmQ0NTQ4NGIyZGQyOGY0MzZmNGU5Yzc4NGNhNDdjODQyMDMyMTgwODc0NGViM2JjNjIyMjVcXD94aXA9M052MFpCSi1hZWcmcmVzcG9uc2UtY29udGVudC1kaXNwb3NpdG

In [ ]:
# /content/GeneratedLabelledFlows.zip

In [ ]:
with zipfile.ZipFile("/content/GeneratedLabelledFlows.zip", "r") as z:
    names = [n for n in z.namelist() if "WebAttacks" in n]
    print(names)
    z.extractall("/content/cicids", members=names)

['TrafficLabelling /Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv']


In [ ]:
!ls -lah /content/cicids/'TrafficLabelling '/Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv

-rw-r--r-- 1 root root 88M Sep 29 20:36 '/content/cicids/TrafficLabelling /Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv'


In [ ]:
old_path = "/content/cicids/TrafficLabelling "
new_path = "/content/cicids/TrafficLabelling"
os.rename(old_path, new_path)

In [ ]:

os.rename(
    f"{new_path}/Thursday-WorkingHours-Morning-WebAttacks.pcap_ISCX.csv",
    f"{new_path}/thursday_web_attacks.csv"
)

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
dataset_path = "/content/cicids/TrafficLabelling/thursday_web_attacks.csv"

In [5]:
df = pd.read_csv(dataset_path,  encoding="latin1")

/tmp/ipykernel_11834/3388758639.py:1: DtypeWarning: Columns (0,1,3,6,84) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(dataset_path,  encoding="latin1")


In [6]:
print(df.shape)
print(df.columns.tolist())

(458968, 85)
['Flow ID', ' Source IP', ' Source Port', ' Destination IP', ' Destination Port', ' Protocol', ' Timestamp', ' Flow Duration', ' Total Fwd Packets', ' Total Backward Packets', 'Total Length of Fwd Packets', ' Total Length of Bwd Packets', ' Fwd Packet Length Max', ' Fwd Packet Length Min', ' Fwd Packet Length Mean', ' Fwd Packet Length Std', 'Bwd Packet Length Max', ' Bwd Packet Length Min', ' Bwd Packet Length Mean', ' Bwd Packet Length Std', 'Flow Bytes/s', ' Flow Packets/s', ' Flow IAT Mean', ' Flow IAT Std', ' Flow IAT Max', ' Flow IAT Min', 'Fwd IAT Total', ' Fwd IAT Mean', ' Fwd IAT Std', ' Fwd IAT Max', ' Fwd IAT Min', 'Bwd IAT Total', ' Bwd IAT Mean', ' Bwd IAT Std', ' Bwd IAT Max', ' Bwd IAT Min', 'Fwd PSH Flags', ' Bwd PSH Flags', ' Fwd URG Flags', ' Bwd URG Flags', ' Fwd Header Length', ' Bwd Header Length', 'Fwd Packets/s', ' Bwd Packets/s', ' Min Packet Length', ' Max Packet Length', ' Packet Length Mean', ' Packet Length Std', ' Packet Length Variance', 'FIN 

In [7]:
df.columns = df.columns.str.strip()

In [8]:
print(df["Label"].value_counts())

Label
BENIGN                        168186
Web Attack – Brute Force        1507
Web Attack – XSS                 652
Web Attack – Sql Injection        21
Name: count, dtype: int64


In [9]:
print("Infinites:", (df.select_dtypes(include="number") == float("inf")).sum().sum())

Infinites: 250


In [10]:
print("NaNs:", df.isna().sum().sum())

NaNs: 24531190


In [11]:
#%%
print(df["Label"].isna().sum())
print(df.isna().sum().sort_values(ascending=False).head(15))

288602
Flow Bytes/s                   288622
Source IP                      288602
Source Port                    288602
Destination IP                 288602
Flow ID                        288602
Destination Port               288602
Protocol                       288602
Flow Duration                  288602
Timestamp                      288602
Total Backward Packets         288602
Total Length of Fwd Packets    288602
Total Length of Bwd Packets    288602
Total Fwd Packets              288602
Fwd Packet Length Max          288602
Fwd Packet Length Min          288602
dtype: int64


In [12]:
df = df.dropna(subset=["Label"]).reset_index(drop=True)

In [13]:
# this dataset is  SOOOO Dirty !!!
print(df.shape)
print(df.isna().sum().sum())
print((df.select_dtypes(include="number") == float("inf")).sum().sum())

(170366, 85)
20
250


In [15]:
# clean all remaining NaN
num_cols = df.select_dtypes(include="number").columns

In [16]:
nan_cols = df[num_cols].isna().sum()
print(nan_cols[nan_cols > 0])

Flow Bytes/s    20
dtype: int64


In [17]:
inf_cols = (df[num_cols] == float("inf")).sum()
print(inf_cols[inf_cols > 0])

Flow Bytes/s      115
Flow Packets/s    135
dtype: int64


In [18]:
## Clean Clean Clean
df["Flow Bytes/s"] = df["Flow Bytes/s"].replace([np.inf, -np.inf], np.nan)
df["Flow Packets/s"] = df["Flow Packets/s"].replace([np.inf, -np.inf], np.nan)

In [19]:
df["Flow Bytes/s"] = df["Flow Bytes/s"].fillna(df["Flow Bytes/s"].median())
df["Flow Packets/s"] = df["Flow Packets/s"].fillna(df["Flow Packets/s"].median())

In [20]:
print(df.isna().sum().sum())

0


In [21]:
print((df.select_dtypes(include="number") == float("inf")).sum().sum())

0


In [22]:
df.shape

(170366, 85)

In [23]:
df.head()

,Flow ID,Source IP,Source Port,Destination IP,Destination Port,Protocol,Timestamp,Flow Duration,Total Fwd Packets,Total Backward Packets,...,min_seg_size_forward,Active Mean,Active Std,Active Max,Active Min,Idle Mean,Idle Std,Idle Max,Idle Min,Label
0,192.168.10.3-192.168.10.50-389-33898-6,192.168.10.50,33898.0,192.168.10.3,389.0,6.0,6/7/2017 8:59,113095465.0,48.0,24.0,...,32.0,203985.500,5.758373e+05,1629110.0,379.0,13800000.0,4.277541e+06,16500000.0,6737603.0,BENIGN
1,192.168.10.3-192.168.10.50-389-33904-6,192.168.10.50,33904.0,192.168.10.3,389.0,6.0,6/7/2017 8:59,113473706.0,68.0,40.0,...,32.0,178326.875,5.034269e+05,1424245.0,325.0,13800000.0,4.229413e+06,16500000.0,6945512.0,BENIGN
2,8.0.6.4-8.6.0.1-0-0-0,8.6.0.1,0.0,8.0.6.4,0.0,0.0,6/7/2017 8:59,119945515.0,150.0,0.0,...,0.0,6909777.333,1.170000e+07,20400000.0,6.0,24400000.0,2.430000e+07,60100000.0,5702188.0,BENIGN
3,192.168.10.14-65.55.44.109-59135-443-6,192.168.10.14,59135.0,65.55.44.109,443.0,6.0,6/7/2017 8:59,60261928.0,9.0,7.0,...,20.0,0.000,0.000000e+00,0.0,0.0,0.0,0.000000e+00,0.0,0.0,BENIGN
4,192.168.10.3-192.168.10.14-53-59555-17,192.168.10.14,59555.0,192.168.10.3,53.0,17.0,6/7/2017 8:59,269.0,2.0,2.0,...,32.0,0.000,0.000000e+00,0.0,0.0,0.0,0.000000e+00,0.0,0.0,BENIGN


There is not Algortihm that can  handle this in-balance classes,  Hence i go with a binary model

BENIGN               168186   (98.7%)

Brute Force            1507   (0.88%)

XSS                     652   (0.38%)

Sql Injection             21   (0.01%)


In [ ]:
df["Label"] = df["Label"].str.strip()

In [ ]:
label_map = {
    "BENIGN": 0,
    "Web Attack  Brute Force": 1,
    "Web Attack  XSS": 1,
    "Web Attack  Sql Injection": 1,
}

In [ ]:
df["target"] = df["Label"].map(label_map)

In [ ]:
print(df["target"].value_counts())


target
0.0    168186
Name: count, dtype: int64


In [ ]:
print(df["target"].isna().sum())

2180


In [ ]:
#%%
for val in df["Label"].unique():
    print(repr(val))

'BENIGN'
'Web Attack \x96 Brute Force'
'Web Attack \x96 XSS'
'Web Attack \x96 Sql Injection'


In [ ]:
'BENIGN'
'Web Attack \x96 Brute Force'
'Web Attack \x96 XSS'
'Web Attack \x96 Sql Injection'


'Web Attack \x96 Sql Injection'

In [24]:
label_map = {
    "BENIGN": 0,
    "Web Attack \x96 Brute Force": 1,
    "Web Attack \x96 XSS": 1,
    "Web Attack \x96 Sql Injection": 1,
}

In [25]:
df["target"] = df["Label"].map(label_map)

In [26]:
print(df["target"].value_counts())
print(df["target"].isna().sum())

target
0    168186
1      2180
Name: count, dtype: int64
0


In [27]:
drop_cols = ["Flow ID", "Source IP", "Destination IP", "Timestamp", "Label", "target"]


In [28]:
X = df.drop(columns=drop_cols)
y = df["target"]

In [29]:
print(X.shape)
print(X.dtypes.value_counts())

(170366, 80)
float64    80
Name: count, dtype: int64


In [30]:
df.columns

Index(['Flow ID', 'Source IP', 'Source Port', 'Destination IP',
       'Destination Port', 'Protocol', 'Timestamp', 'Flow Duration',
       'Total Fwd Packets', 'Total Backward Packets',
       'Total Length of Fwd Packets', 'Total Length of Bwd Packets',
       'Fwd Packet Length Max', 'Fwd Packet Length Min',
       'Fwd Packet Length Mean', 'Fwd Packet Length Std',
       'Bwd Packet Length Max', 'Bwd Packet Length Min',
       'Bwd Packet Length Mean', 'Bwd Packet Length Std', 'Flow Bytes/s',
       'Flow Packets/s', 'Flow IAT Mean', 'Flow IAT Std', 'Flow IAT Max',
       'Flow IAT Min', 'Fwd IAT Total', 'Fwd IAT Mean', 'Fwd IAT Std',
       'Fwd IAT Max', 'Fwd IAT Min', 'Bwd IAT Total', 'Bwd IAT Mean',
       'Bwd IAT Std', 'Bwd IAT Max', 'Bwd IAT Min', 'Fwd PSH Flags',
       'Bwd PSH Flags', 'Fwd URG Flags', 'Bwd URG Flags', 'Fwd Header Length',
       'Bwd Header Length', 'Fwd Packets/s', 'Bwd Packets/s',
       'Min Packet Length', 'Max Packet Length', 'Packet Length Mean',
  

In [31]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, roc_auc_score

In [32]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)


In [33]:
scaler = StandardScaler()

In [34]:
X_train_scaled = scaler.fit_transform(X_train)

In [35]:
X_test_scaled = scaler.transform(X_test)

In [36]:
model = LogisticRegression(class_weight="balanced", max_iter=1000, random_state=42)

In [37]:
model.fit(X_train_scaled, y_train)

LogisticRegression(class_weight='balanced', max_iter=1000, random_state=42)

In [38]:
y_pred = model.predict(X_test_scaled)

In [39]:
y_proba = model.predict_proba(X_test_scaled)[:, 1]

In [40]:
print(classification_report(y_test, y_pred, digits=4))

              precision    recall  f1-score   support

           0     0.9999    0.9784    0.9890     33638
           1     0.3735    0.9954    0.5432       436

    accuracy                         0.9786     34074
   macro avg     0.6867    0.9869    0.7661     34074
weighted avg     0.9919    0.9786    0.9833     34074



In [41]:
print("ROC AUC:", roc_auc_score(y_test, y_proba))

ROC AUC: 0.9976162825899717


In [42]:
from sklearn.metrics import precision_recall_curve

In [43]:
precisions, recalls, thresholds = precision_recall_curve(y_test, y_proba)

In [45]:
target_recall = 0.90
mask = recalls[:-1] >= target_recall
best_idx = np.where(mask)[0][-1]

In [46]:
print(f"threshold={thresholds[best_idx]:.4f}")
print(f"precision={precisions[best_idx]:.4f}")
print(f"recall={recalls[best_idx]:.4f}")

threshold=0.9905
precision=0.7602
recall=0.9014


In [48]:
import joblib

In [49]:
joblib.dump(model, "netflow_lr_model.joblib")
joblib.dump(scaler, "netflow_lr_scaler.joblib")

['netflow_lr_scaler.joblib']

In [50]:
metadata = {
    "model": "netflow_web_attack_lr",
    "classes": {"0": "BENIGN", "1": "Web_Attack"},
    "decision_threshold": 0.9905,
    "source": "CICFlowMeter live capture (TCP/flow-level features)",
    "sklearn_version": "1.8.0",
    "metrics_at_threshold": {
        "precision": 0.7602,
        "recall": 0.9014
    },
    "metrics_at_0.5": {
        "precision": 0.3735,
        "recall": 0.9954,
        "roc_auc": 0.9976162825899717
    }
}

In [52]:
import json

In [54]:
## Save model metadata for upcoming production
with open("netflow_lr_metadata.json", "w") as f:
    json.dump(metadata, f, indent=2)

In [56]:
!ls -lah /content/

total 271M
drwxr-xr-x 1 root root 4.0K Sep 29 21:25 .
drwxr-xr-x 1 root root 4.0K Sep 29 20:25 ..
drwxr-xr-x 3 root root 4.0K Sep 29 20:39 cicids
drwxr-xr-x 4 root root 4.0K Sep 16 13:26 .config
-rw-r--r-- 1 root root 271M Sep 29 20:35 GeneratedLabelledFlows.zip
-rw-r--r-- 1 root root  415 Sep 29 21:25 netflow_lr_metadata.json
-rw-r--r-- 1 root root 1.5K Sep 29 21:24 netflow_lr_model.joblib
-rw-r--r-- 1 root root 4.2K Sep 29 21:24 netflow_lr_scaler.joblib
drwxr-xr-x 1 root root 4.0K Sep 16 13:26 sample_data
